shishua-half-avx2.h

```diff
@@ -0,0 +1,112 @@
+// AVX2 x86_64 implementation of shishua half
+#ifndef SHISHUA_HALF_AVX2_H
+#define SHISHUA_HALF_AVX2_H
+#include <stdint.h>
+#include <stddef.h>
+#include <string.h>
+#include <immintrin.h>
+#include <assert.h>
+
+typedef struct prng_state {
+  __m256i state[2];
+  __m256i output;
+  __m256i counter;
+} prng_state;
+
+// buf could technically alias with prng_state, according to the compiler.
+#if defined(__GNUC__) || defined(_MSC_VER)
+#  define SHISHUA_RESTRICT __restrict
+#else
+#  define SHISHUA_RESTRICT
+#endif
+
+// buf's size must be a multiple of 32 bytes.
+static inline void prng_gen(prng_state *SHISHUA_RESTRICT s, uint8_t *SHISHUA_RESTRICT buf, size_t size) {
+  __m256i s0 = s->state[0], counter = s->counter,
+          s1 = s->state[1],       o = s->output,
+          t0, t1, t2, t3, u0, u1, u2, u3;
+  // The following shuffles move weak (low-diffusion) 32-bit parts of 64-bit
+  // additions to strong positions for enrichment. The low 32-bit part of a
+  // 64-bit chunk never moves to the same 64-bit chunk as its high part.
+  // They do not remain in the same chunk. Each part eventually reaches all
+  // positions ringwise: A to B, B to C, …, H to A.
+  // You may notice that they are simply 256-bit rotations (96 and 160).
+  __m256i shu0 = _mm256_set_epi32(4, 3, 2, 1, 0, 7, 6, 5),
+          shu1 = _mm256_set_epi32(2, 1, 0, 7, 6, 5, 4, 3);
+  // The counter is not necessary to beat PractRand.
+  // It sets a lower bound of 2^69 bytes = 512 EiB to the period,
+  // or about 1 millenia at 10 GiB/s.
+  // The increments are picked as odd numbers,
+  // since only coprimes of the base cover the full cycle,
+  // and all odd numbers are coprime of 2.
+  // I use different odd numbers for each 64-bit chunk
+  // for a tiny amount of variation stirring.
+  // I used the smallest odd numbers to avoid having a magic number.
+  __m256i increment = _mm256_set_epi64x(1, 3, 5, 7);
+<<<<<<< Updated upstream
+  for (size_t i = 0; i < size; i += 32) {
+    _mm256_storeu_si256((__m256i*)&buf[i], o);
+=======
+
+  // TODO: consider adding proper uneven write handling
+  assert((size % 32 == 0) && "buf's size must be a multiple of 32 bytes.");
+
+  for (size_t i = 0; i < size; i += 32) {
+    if (buf != NULL) {
+      _mm256_storeu_si256((__m256i*)&buf[i], o);
+    }
+>>>>>>> Stashed changes
```

Comment:
Forgotten merge conflict I think.

Reply:
Yes, I was a little overtired when I committed this. I will fix it soon.